File: src/wrappers/python/tests/test_PS16.py
import unittest
import string
import logging
from _groupsig import ffi

from pygroupsig import groupsig
from pygroupsig import grpkey
from pygroupsig import mgrkey
from pygroupsig import memkey
from pygroupsig import identity
from pygroupsig import message
from pygroupsig import signature
from pygroupsig import gml
from pygroupsig import constants

# Tests for group operations
class TestGroupOps(unittest.TestCase):

    # Non-test functions
    def addMember(self):
        msg1 = groupsig.join_mgr(0, self.mgrkey, self.grpkey, gml = self.gml)
        msg2 = groupsig.join_mem(1, self.grpkey, msgin = msg1)
        usk = msg2['memkey']
        msg3 = groupsig.join_mgr(2, self.mgrkey, self.grpkey, msg2['msgout'], gml = self.gml)
        msg4 = groupsig.join_mem(3, self.grpkey, msgin = msg3, memkey = usk)
        usk = msg4['memkey']
        self.memkeys.append(usk)

    def setUp(self):
        groupsig.init(constants.PS16_CODE, 0)
        group = groupsig.setup(constants.PS16_CODE)
        self.code = constants.PS16_CODE
        self.mgrkey = group['mgrkey']
        self.grpkey = group['grpkey']
        self.gml = group['gml']
        self.memkeys = []
        
    def tearDown(self):
        groupsig.clear(self.code)

    # Creates a group
    def test_groupCreate(self):
        self.assertNotEqual(self.grpkey, ffi.NULL)
        self.assertNotEqual(self.mgrkey, ffi.NULL)
        self.assertEqual(groupsig.get_joinseq(constants.PS16_CODE), 3)
        self.assertEqual(groupsig.get_joinstart(constants.PS16_CODE), 0)   

    # Adds one member
    def test_addMember(self):
        n_members = len(self.memkeys)
        self.addMember()
        self.assertEqual(len(self.memkeys), n_members+1)
        self.assertNotEqual(self.memkeys[n_members], ffi.NULL)
        
    # Accepts a valid signature for a message passed as a string
    def test_acceptValidSignatureString(self):
        self.addMember()
        sig = groupsig.sign("Hello, World!", self.memkeys[0], self.grpkey)
        b = groupsig.verify(sig, "Hello, World!", self.grpkey)
        self.assertTrue(b)

    # Rejects a valid signature for a different message, also passed as a string
    def test_rejectValidSignatureWrongMessageString(self):
        self.addMember()
        sig = groupsig.sign("Hello, World!", self.memkeys[0], self.grpkey)
        b = groupsig.verify(sig, "Hello, Worlds!", self.grpkey)
        self.assertFalse(b)

    # Accepts a valid signature for a message passed as a byte array
    def test_acceptValidSignatureBytes(self):
        self.addMember()
        sig = groupsig.sign(b"Hello, World!", self.memkeys[0], self.grpkey)
        b = groupsig.verify(sig, b"Hello, World!", self.grpkey)
        self.assertTrue(b)

    # Rejects a valid signature for a different message, also passed as a byte array
    def test_rejectValidSignatureWrongMessageBytes(self):
        self.addMember()
        sig = groupsig.sign(b"Hello, World!", self.memkeys[0], self.grpkey)
        b = groupsig.verify(sig, b"Hello, Worlds!", self.grpkey)
        self.assertFalse(b)

    # Successfully opens a signature
    def test_openSignature(self):
        self.addMember()
        self.addMember()
        sig = groupsig.sign(b"Hello, World!", self.memkeys[1], self.grpkey)
        gsopen = groupsig.open(sig, self.mgrkey, self.grpkey, gml = self.gml)
        self.assertEqual(gsopen["index"], 1)
        proof = gsopen['proof']
        b = groupsig.open_verify(proof, sig, self.grpkey)
        self.assertTrue(b)
        
# Tests for signature operations
class TestSignatureOps(unittest.TestCase):

    # Non-test functions
    def addMember(self):
        msg1 = groupsig.join_mgr(0, self.mgrkey, self.grpkey, gml = self.gml)
        msg2 = groupsig.join_mem(1, self.grpkey, msgin = msg1)
        usk = msg2['memkey']
        msg3 = groupsig.join_mgr(2, self.mgrkey, self.grpkey, msg2['msgout'], gml = self.gml)
        msg4 = groupsig.join_mem(3, self.grpkey, msgin = msg3, memkey = usk)
        usk = msg4['memkey']
        self.memkeys.append(usk)
        
    # Creates a group, adds a member and generates a signature
    def setUp(self):
        groupsig.init(constants.PS16_CODE, 0)
        group = groupsig.setup(constants.PS16_CODE)
        self.code = constants.PS16_CODE
        self.mgrkey = group['mgrkey']
        self.grpkey = group['grpkey']
        self.gml = group['gml']
        self.memkeys = []
        self.addMember()
        self.sig = groupsig.sign("Hello, World!", self.memkeys[0], self.grpkey)
        
    def tearDown(self):
        groupsig.clear(self.code)

    # Exports and reimports a signature, and it verifies correctly
    def test_sigExportImport(self):
        sig_str = signature.signature_export(self.sig)
        sig = signature.signature_import(self.code, sig_str)
        b = groupsig.verify(sig, "Hello, World!", self.grpkey)
        self.assertTrue(b)
                        
    # Prints a string (this just checks the produced string is not empty)
    def test_sigToString(self):
        sig_str = signature.signature_to_string(self.sig)
        self.assertGreater(len(sig_str), 0)
        self.assertTrue(set(sig_str).issubset(set(string.printable)))

# Tests for group key operations
class TestGrpkeyOps(unittest.TestCase):

    # Creates a group, adds a member and generates a signature
    def setUp(self):
        groupsig.init(constants.PS16_CODE, 0)
        group = groupsig.setup(constants.PS16_CODE)
        self.code = constants.PS16_CODE
        self.mgrkey = group['mgrkey']
        self.grpkey = group['grpkey']
        self.gml = group['gml']
        
    def tearDown(self):
        groupsig.clear(self.code)

    # Exports and reimports a group key
    def test_grpkeyExportImport(self):
        grpkey_str = grpkey.grpkey_export(self.grpkey)
        gpk = grpkey.grpkey_import(self.code, grpkey_str)
        # This is quite useless, as import returns an exception if the FFI
        # method returns ffi.NULL. Maybe implementing a cmp function for
        # grp keys would be good for testing this (and also in general?)
        self.assertIsNot(ffi.NULL, gpk)

# Tests for manager key operations
class TestManagerkeyOps(unittest.TestCase):

    # Creates a group, adds a member and generates a signature
    def setUp(self):
        groupsig.init(constants.PS16_CODE, 0)
        group = groupsig.setup(constants.PS16_CODE)
        self.code = constants.PS16_CODE
        self.mgrkey = group['mgrkey']
        self.grpkey = group['grpkey']
        self.gml = group['gml']
        
    def tearDown(self):
        groupsig.clear(self.code)

    # Exports and reimports an manager key
    def test_mgrkeyExportImport(self):
        mgrkey_str = mgrkey.mgrkey_export(self.mgrkey)
        ikey = mgrkey.mgrkey_import(self.code, mgrkey_str)
        # This is quite useless, as import returns an exception if the FFI
        # method returns ffi.NULL. Maybe implementing a cmp function for
        # manager keys would be good for testing this (and also in general?)
        self.assertIsNot(ffi.NULL, ikey)

# Tests for member key operations
class TestMemkeyOps(unittest.TestCase):

    # Non-test functions
    def addMember(self):
        msg1 = groupsig.join_mgr(0, self.mgrkey, self.grpkey, gml = self.gml)
        msg2 = groupsig.join_mem(1, self.grpkey, msgin = msg1)
        usk = msg2['memkey']
        msg3 = groupsig.join_mgr(2, self.mgrkey, self.grpkey, msg2['msgout'], gml = self.gml)
        msg4 = groupsig.join_mem(3, self.grpkey, msgin = msg3, memkey = usk)
        usk = msg4['memkey']
        self.memkey = usk
        
    # Creates a group, adds a member and generates a signature
    def setUp(self):
        groupsig.init(constants.PS16_CODE, 0)
        group = groupsig.setup(constants.PS16_CODE)
        self.code = constants.PS16_CODE
        self.mgrkey = group['mgrkey']
        self.grpkey = group['grpkey']
        self.gml = group['gml']
        self.addMember()
        
    def tearDown(self):
        groupsig.clear(self.code)

    # Exports and reimports a member key
    def test_memkeyExportImport(self):
        memkey_str = memkey.memkey_export(self.memkey)
        mkey = memkey.memkey_import(self.code, memkey_str)
        # This is quite useless, as import returns an exception if the FFI
        # method returns ffi.NULL. Maybe implementing a cmp function for
        # mem keys would be good for testing this (and also in general?)
        self.assertIsNot(ffi.NULL, mkey)

# Tests for GML operations
class TestGmlOps(unittest.TestCase):

    # Non-test functions
    def addMember(self):
        msg1 = groupsig.join_mgr(0, self.mgrkey, self.grpkey, gml = self.gml)
        msg2 = groupsig.join_mem(1, self.grpkey, msgin = msg1)
        usk = msg2['memkey']
        msg3 = groupsig.join_mgr(2, self.mgrkey, self.grpkey, msg2['msgout'], gml = self.gml)
        msg4 = groupsig.join_mem(3, self.grpkey, msgin = msg3, memkey = usk)
        usk = msg4['memkey']
        self.memkey = usk
        
    # Creates a group, adds a member and generates a signature
    def setUp(self):
        groupsig.init(constants.PS16_CODE, 0)
        group = groupsig.setup(constants.PS16_CODE)
        self.code = constants.PS16_CODE
        self.mgrkey = group['mgrkey']
        self.grpkey = group['grpkey']
        self.gml = group['gml']
        self.addMember()
        
    def tearDown(self):
        groupsig.clear(self.code)

    # Exports and reimports a member key
    def test_gmlExportImport(self):
        gml_str = gml.gml_export(self.gml)
        _gml = gml.gml_import(self.code, gml_str)
        # This is quite useless, as import returns an exception if the FFI
        # method returns ffi.NULL. Maybe implementing a cmp function for
        # GMLs would be good for testing this (and also in general?)
        self.assertIsNot(ffi.NULL, _gml)
                                
# Define test suites
def suiteGroupOps():
    suiteGroupOps = unittest.TestSuite()    
    suiteGroupOps.addTest(WidgetTestCase('test_groupCreate'))
    suiteGroupOps.addTest(WidgetTestCase('test_addMember'))
    suiteGroupOps.addTest(WidgetTestCase('test_acceptValidSignatureString'))
    suiteGroupOps.addTest(WidgetTestCase('test_rejectValidSignatureWrongMessageString'))
    suiteGroupOps.addTest(WidgetTestCase('test_acceptValidSignatureBytes'))
    suiteGroupOps.addTest(WidgetTestCase('test_rejectValidSignatureWrongMessageBytes'))
    suiteGroupOps.addTest(WidgetTestCase('test_openSignature'))
    return suiteGroupOps
        
def suiteSigOps():
    suiteSigOps = unittest.TestSuite()    
    suiteSigOps.addTest(WidgetTestCase('test_sigExportImport'))
    suiteSigOps.addTest(WidgetTestCase('test_sigToString'))
    return suiteSigOps

def suiteGrpkeyOps():
    suiteGrpkeyOps = unittest.TestSuite()    
    suiteGrpkeyOps.addTest(WidgetTestCase('test_grpkeyExportImport'))
    return suiteGrpkeyOps

def suiteManagerkeyOps():
    suiteManagerkeyOps = unittest.TestSuite()    
    suiteManagerkeyOps.addTest(WidgetTestCase('test_mgrkeyExportImport'))
    return suiteManagerkeyOps

def suiteMemkeyOps():
    suiteMemkeyOps = unittest.TestSuite()    
    suiteMemkeyOps.addTest(WidgetTestCase('test_memkeyExportImport'))
    return suiteMemkeyOps

def suiteGmlOps():
    suiteGmlOps = unittest.TestSuite()    
    suiteGmlOps.addTest(WidgetTestCase('test_gmlExportImport'))
    return suiteGmlOps
        
if __name__ == '__main__':
    runner = unittest.TextTestRunner()
    runner.run(suiteGroupOps())
    runner.run(suiteSigOps())
    runner.run(suiteGrpkeyOps())
    runner.run(suiteManagerkeyOps())
    runner.run(suiteMemkeyOps())
    runner.run(suiteGmlOps())

File: src/wrappers/nodejs/jsgroupsig/test/test_PS16.js
let chai = require('chai');
const expect = require('chai').expect;
let assert = require('assert');

const jsgroupsig = require('../lib/index.js');

function setupFull() {
    /* Note: for testing we run all the setup calls within this function.
       In the real world, the first call to setup should be done by the issuer,
       the second by the converter. See the repository documentation for more
       info. */
    jsgroupsig.init(jsgroupsig.PS16);
    let ps16 = jsgroupsig.get_groupsig_from_code(jsgroupsig.PS16);
    let grpkey = jsgroupsig.grp_key_init(jsgroupsig.PS16);
    let mgrkey = jsgroupsig.mgr_key_init(jsgroupsig.PS16);
    let gml = jsgroupsig.gml_init(jsgroupsig.PS16);
    jsgroupsig.setup(jsgroupsig.PS16, grpkey, mgrkey, gml);
    return {
	'grpkey': grpkey,
	'mgrkey': mgrkey,
	'gml': gml
    };
}

function addMember(mgrkey, grpkey, gml) {
    let memkey = jsgroupsig.mem_key_init(jsgroupsig.PS16);
    let msg1 = jsgroupsig.join_mgr(0, mgrkey, grpkey, null, gml);
    let msg2 = jsgroupsig.join_mem(1, memkey, grpkey, msg1);
    let msg3 = jsgroupsig.join_mgr(2, mgrkey, grpkey, msg2, gml);
    jsgroupsig.join_mem(3, memkey, grpkey, msg3);
    
    return memkey;
}

describe('PS16 Group key operations', function() {

    it('initializes a group key.', function() {
	jsgroupsig.init(jsgroupsig.PS16, 0);	
	let grpkey = jsgroupsig.grp_key_init(jsgroupsig.PS16);
	assert.notEqual(grpkey, null);
    });

    it('exports and imports a group key.', function() {
	let ps16 = setupFull();
	let str = jsgroupsig.grp_key_export(ps16.grpkey);
	let grpkey = jsgroupsig.grp_key_import(jsgroupsig.PS16, str);
	assert.notEqual(grpkey, null);
    });

});

describe('PS16 Manager key operations', function() {

    it('initializes a manager key.', function() {
	jsgroupsig.init(jsgroupsig.PS16, 0);	
	let mgrkey = jsgroupsig.mgr_key_init(jsgroupsig.PS16);
	assert.notEqual(mgrkey, null);
    });

    it('exports and imports a manager key.', function() {
	let ps16 = setupFull();
	let str = jsgroupsig.mgr_key_export(ps16.mgrkey);
	let mgrkey = jsgroupsig.mgr_key_import(jsgroupsig.PS16, str);
	assert.notEqual(mgrkey, null);
    });


});

describe('PS16 Member key operations', function() {

    it('initializes a member key.', function() {
	jsgroupsig.init(jsgroupsig.PS16, 0);	
	let memkey = jsgroupsig.mem_key_init(jsgroupsig.PS16);
	assert.notEqual(memkey, null);
    });

    it('exports and imports a member key.', function() {
	let ps16 = setupFull();
	let memkey = addMember(ps16.mgrkey, ps16.grpkey, ps16.gml);
	let str = jsgroupsig.mem_key_export(memkey);
	let memkey2 = jsgroupsig.mem_key_import(jsgroupsig.PS16, str);
	assert.notEqual(memkey2, null);
    });    

});

describe('PS16 Signature operations', function() {

    it('initializes a signature.', function() {
	jsgroupsig.init(jsgroupsig.PS16, 0);
	let sig = jsgroupsig.signature_init(jsgroupsig.PS16);
	assert.notEqual(sig, null);
    });

    it('exports and imports a signature.', function() {
	let ps16 = setupFull();
	let memkey = addMember(ps16.mgrkey, ps16.grpkey, ps16.gml);
	let sig = jsgroupsig.sign("Hello, World!", memkey, ps16.grpkey);
	let str = jsgroupsig.signature_export(sig);
	let sig2 = jsgroupsig.signature_import(jsgroupsig.PS16, str);
	let b = jsgroupsig.verify(sig2, "Hello, World!", ps16.grpkey);
	assert.equal(b, true);
    });      
    
});

describe('PS16 Group operations', function() {
    
    it('sets up group and manager keys.', function() {
	let ps16 = setupFull();
	assert.notEqual(ps16.grpkey, null);
	assert.notEqual(ps16.mgrkey, null);
    });

    it('the manager starts the join protocol', function() {
	let s = jsgroupsig.get_joinstart(jsgroupsig.PS16);
	assert.equal(s, 0);
    });

    it('the join protocol has 4 messages', function() {
	let s = jsgroupsig.get_joinseq(jsgroupsig.PS16);
	assert.equal(s, 3); // start counting in 0
    });

    it('adds a new group member.', function() {
	let ps16 = setupFull();
	let memkey = addMember(ps16.mgrkey, ps16.grpkey, ps16.gml);
	assert.notEqual(memkey, null);
    });

    it('a VALID string signature is accepted.', function() {
	let ps16 = setupFull();
	let memkey = addMember(ps16.mgrkey, ps16.grpkey, ps16.gml);
	let sig = jsgroupsig.sign("Hello, World!", memkey, ps16.grpkey);
	let b = jsgroupsig.verify(sig, "Hello, World!", ps16.grpkey);
	assert.equal(b, true);
    });

    it('a WRONG string signature is rejected.', function() {
	let ps16 = setupFull();
	let memkey = addMember(ps16.mgrkey, ps16.grpkey, ps16.gml);
	let sig = jsgroupsig.sign("Hello, World!", memkey, ps16.grpkey);
	let b = jsgroupsig.verify(sig, "Hello, World2!", ps16.grpkey);
	assert.equal(b, false);
    });

    it('correctly opens a signature.', function() {
	let ps16 = setupFull();
	let memkey = addMember(ps16.mgrkey, ps16.grpkey, ps16.gml);
	let memkey2 = addMember(ps16.mgrkey, ps16.grpkey, ps16.gml);
	let sig = jsgroupsig.sign("Hello, World!", memkey2, ps16.grpkey);
	let { index, proof } = jsgroupsig.open(sig, ps16.grpkey, ps16.mgrkey, ps16.gml);
	assert.equal(index, 1);
	let b = jsgroupsig.open_verify(proof, sig, ps16.grpkey);
	assert.equal(b, true);
    });

});

